ff1/ff1.go

```diff
@@ -121,106 +126,119 @@ func (f Cipher) Encrypt(X string) (string, error) {
 		return ret, errors.New("message length is not within min and max bounds")
 	}
 
-	// Check if the message is in the current radix by using the numRadix function
-	_, err = numRadix(X, f.radix)
-	if err != nil {
+	radix := f.radix
+
+	// Check if the message is in the current radix by using the same logic as numRadix
+	var bX big.Int
+	_, ok := bX.SetString(X, radix)
+	if !ok {
 		return ret, errors.New("message is not within base/radix")
 	}
 
 	// Calculate split point
-	u := uint32(math.Floor(float64(n) / 2))
+	u := n / 2
 	v := n - u
 
 	// Split the message
 	A := X[:u]
 	B := X[u:]
 
 	// Byte lengths
-	b := int(math.Ceil(math.Ceil(float64(v)*math.Log2(float64(f.radix))) / 8))
+	b := int(math.Ceil(math.Ceil(float64(v)*math.Log2(float64(radix))) / 8))
+
+	// TODO: why does this cause cumulous allocations?
 	d := int(4*math.Ceil(float64(b)/4) + 4)
 
-	// Calculate P
-	P := bytes.NewBuffer([]byte{})
-	_, err = P.Write([]byte{0x01, 0x02, 0x01, 0x00})
-	if err != nil {
-		return ret, err
-	}
+	// Calculate P, which is always 16 bytes
+	P := make([]byte, 16)
+	P[0] = 0x01
+	P[1] = 0x02
+	P[2] = 0x01
 
-	err = binary.Write(P, binary.BigEndian, uint16(f.radix))
-	if err != nil {
-		return ret, err
-	}
+	// radix must fill 3 bytes, so pad 1 zero byte
+	P[3] = 0x00
+	binary.BigEndian.PutUint16(P[4:6], uint16(radix))
 
-	err = P.WriteByte(0x0a)
-	if err != nil {
-		return ret, err
-	}
+	P[6] = 0x0a
+	P[7] = byte(u) // overflow automatically does the modulus
 
-	err = P.WriteByte(uint8(u % 256))
-	if err != nil {
-		return ret, err
-	}
+	binary.BigEndian.PutUint32(P[8:12], n)
 
-	err = binary.Write(P, binary.BigEndian, n)
-	if err != nil {
-		return ret, err
+	binary.BigEndian.PutUint32(P[12:16], t)
+
+	// These are re-used in the for loop below
+	var (
+		// Q's length is a multiple of 16
+		// Start it with a length of 16
+		Q     = make([]byte, 16)
+		QOrig = make([]byte, 0, 16)
+
+		// R is gauranteed to be 16 bytes since it holds output of PRF
+		R = make([]byte, 16)
+
+		// TODO: rename this
+		temp []byte
+
+		numB      big.Int
+		numBBytes []byte
+
+		br, bm, mod big.Int
+		y, c        big.Int
+
+		Y []byte
+	)
+
+	br.SetInt64(int64(radix))
+
+	maxJ := int(math.Ceil(float64(d) / 16))
+
+	// This is the fixed part of Q
+	numPad := (-int(t) - b - 1) % 16
+	if numPad < 0 {
+		numPad += 16
 	}
 
-	err = binary.Write(P, binary.BigEndian, t)
-	if err != nil {
-		return ret, err
+	QOrig = append(QOrig, f.tweak...)
+	QOrig = append(QOrig, ivZero[:numPad]...)
+
+	// PQ is at least 32 bytes, and the firs part is always P
+	// PQ := make([]byte, 0, 32)
+	// PQ = append(PQ, P...)
+
+	// temp must be 16 bytes incuding j
+	// This will only be needed if maxJ > 1, for the inner for loop
+	if maxJ > 1 {
+		temp = make([]byte, 16)
```

Comment:
This is a good suggestion, thanks! I was initially a little apprehensive because I was worried about readability but the variables remain as is, it's just the underlying backing array that would change.